dep0/lib/03_typecheck/src/check.cpp

```diff
@@ -65,142 +64,195 @@ stmt_t make_legal_stmt(Args&&... args)
+        pretty_print(err << "cannot invoke expression of type `", func->properties.sort) << '`';
+        return error(err.str());
+    }();
+    if (not func_type)
+        return func_type.error();
+    if (func_type->arg_types.size() != f.args.size())
     {
         std::ostringstream err;
-        err << "passed " << f.args.size() << " arguments but was expecting " << proto->value.args.size();
-        return error({dep0::error_t{err.str()}});
+        err << "passed " << f.args.size() << " arguments but was expecting " << func_type->arg_types.size();
+        return error(err.str());
     }
     std::vector<expr_t> args;
-    std::vector<dep0::error_t> reasons;
-    for (auto const i: std::views::iota(0ul, f.args.size()))
-        if (auto expr = check(ctx, f.args[i], proto->value.args[i].type))
-            args.push_back(std::move(*expr));
+    for (auto const i: std::views::iota(0ul, func_type->arg_types.size()))
+    {
+        auto arg = match(
+            func_type->arg_types[i],
+            [&] (type_t::var_t const& var) -> expected<expr_t>
+            {
+                if (auto type = check_typename(ctx, f.args[i], ast::typename_t{}))
+                {
+                    // This might be worth an explanation. In "classical" Type Theory we cannot blindly perform
+                    // substitution in the return type. Instead we should check whether a new type-variable with
+                    // the same name as `var` was introduced by some later argument.
+                    // If it was, we cannot perform substitution in the return type because now the `var` in the
+                    // return type would be bound to the other `var`, not this one.
+                    // However, when typechecking a function prototype, we disallow shadowing between arguments
+                    // at "the same level", so we know that the return type can only bind to this `var`.
+                    // More concretely:
+                    //  - `auto f(typename t, typename t) -> t` would, in theory, be ok and the return type
+                    //    would be the `t` from the second argument, but we disallow this kind of shadowing;
+                    //  - this is different from `auto g(typename t, (typename t) -> t) -> t` because the return
+                    //    type is bound to the `t` in the first argument, not the second one; we allow this.
+                    std::for_each(
+                        func_type->arg_types.begin() + i + 1,
+                        func_type->arg_types.end(),
+                        [&] (type_t::arr_t::arg_type_t& x)
+                        {
+                            match(
+                                x,
+                                [&] (type_t::var_t const& v)
+                                {
+                                    assert(v != var); // because of the comment above
+                                    // TODO rename `v` if it appears as free variable in `y`
```

Comment:
I should try with this test
```
int zero() { return 0; }
auto f(typename t, typename int, t x) -> t { return x; }
f(int, i32_t, zero()) : int

Reply:
Yes, this works, as in it fails to typecheck, in fact this can be reduced to
```
auto f(typename t, typename int, t x) -> t { return x; }
int g() { return f(int, i32_t, 0); }
```
changing `int` to `int_` makes it typecheck